src/proxy/pool.rs

```diff
@@ -12,80 +12,398 @@
+            //
+            // critical block - this writelocks the entire pool for all tasks/threads
+            // as we check to see if anyone has ever inserted a sharded mutex for this key.
+            //
+            // If not, we insert one.
+            //
+            // We want to hold this for as short as possible a time, and drop it
+            // before we hold it over an await.
+            //
+            // this is is the ONLY block where we should hold a writelock on the whole mutex map
+            // for the rest, a readlock (nonexclusive) is sufficient.
+            {
+                let mut map_write_lock = self.established_conn_writelock.write().await;
+                match map_write_lock.get(&hash_key) {
+                    Some(_) => {
+                        debug!("already have conn for key {:#?}", hash_key);
+                    }
+                    None => {
+                        debug!("inserting conn mutex for key {:#?}", hash_key);
+                        map_write_lock.insert(hash_key, Some(Mutex::new(())));
+                    }
+                };
+                drop(map_write_lock); // strictly redundant
+            }
+
+            // If we get here, it means the following are true:
+            // 1. We have a guaranteed sharded mutex in the outer map for our current key.
+            // 2. We can now, under readlock(nonexclusive) in the outer map, attempt to
+            // take the inner writelock for our specific key (exclusive).
+            //
+            // This unblocks other tasks spawning connections against other keys, but blocks other
+            // tasks spawning connections against THIS key - which is what we want.
+
+            // NOTE: This inner, key-specific mutex is a tokio::async::Mutex, and not a stdlib sync mutex.
+            // these differ from the stdlib sync mutex in that they are (slightly) slower
+            // (they effectively sleep the current task) and they can be held over an await.
+            // The tokio docs (rightly) advise you to not use these,
+            // because holding a lock over an await is a great way to create deadlocks if the await you
+            // hold it over does not resolve.
+            //
+            // HOWEVER. Here we know this connection will either establish or timeout
+            // and we WANT other tasks to go back to sleep if there is an outstanding lock.
+            // So the downsides are actually useful (we WANT task contention -
+            // to block other parallel tasks from trying to spawn a connection if we are already doing so)
+            //
+            // BEGIN take outer readlock
+            let map_read_lock = self.established_conn_writelock.read().await;
+            let exist_conn_lock = map_read_lock.get(&hash_key).unwrap();
+            // BEGIN take inner writelock
+            let found_conn = match exist_conn_lock.as_ref().unwrap().try_lock() {
```

Comment:
The locking seems problematic. Talking about the outer lock here.

1. We open connection with readlock. Connection can take ~20s
2. New connection starts with different key. It tries to writelock. This blocks until all readers  (1) are done.
3. New connection starts for an existing key. This tries to readlock. This blocks until writer (2) are done.

End result: all new connections blocked on one connection
?

Reply:
Yes, this will block on net-new connections. It should not on reusing established connections.

If a workload opens many connections to many different endpoints at once, it will effectively be throttled by this.

The only alternative to a pool-wide state I can think of is race-discard, which throttles everybody, including neighbors - this just throttles the workload.

Reply:
In the scenario about even entirely different pool keys can block.

A -> B can block A -> C. B and C could be different nodes entirely as well

Reply:
Taking a step back, here is what I see as the requirements

There should be a Pool, per proxy instance (pod, for inpod mode)

It should expose, something like
`  fn send_request(key, Request) -> Response`

With the following behavior:
If there is an existing connection open with available streams for the key, send the request
   Because hyper doesn't expose active streams and max streams, this will need to
   increment a per-connection "active streams" counter at the start, and make a Drop impl decrement.
Else if there is an existing connection pending for the key, and the number of pending+active requests is < max streams, wait for it to complete 
Else, establish a new connection and put it in the pool

After the last request on a connection is finished, it is marked idle. Connections idle for more than ~90s should be closed

State of connections for other keys than the one requested should not impact

Not strictly required, but a good design sign: No IO should be done with locks held

---

Hyper's pool falls short in two ways here:
* We don't have a way to not re-use an existing steam because `reserve()` is infallible. This could probably be fixed with a `try_reserve()` extension (upstream)
* We don't have a way to say not to re-use a connect because its going to be full when we get it

Reply:
I am not sure I follow why we need to chose between "open excessive connections" and "unrelated keys can block each other"?

Do you mean even in abstract these are in contention, or its challenging to code it that way?

I think it is conceptually possible, though all of the pooling logic is tricky (this, pingora, hyper-util -- the general concept, really)

Reply:
A third way it falls short:

- We don't have a natural way to manage/track more than one active connection in a hyper pool to the same dest.

> If there is an existing connection open with available streams for the key, send the request

> Because hyper doesn't expose active streams and max streams, this will need to
> increment a per-connection "active streams" counter at the start, and make a Drop impl decrement.

If we remove/deschedule full streams from the pool, the decrement is probably not required, but yes.

> Else if there is an existing connection pending for the key, and the number of pending+active requests is < max streams, wait for it to complete

Yes, the key here being that we need to synchronize the spawning of a new connection, such that at-most-one task "wins" the right to spawn - especially in scenarios where 100s of tasks might spin  up trying to do this.

> Else, establish a new connection and put it in the pool

Yep.

> After the last request on a connection is finished, it is marked idle. Connections idle for more than ~90s should be closed

Yep, that's a todo here. This doesn't strictly have to be managed by the pool though, it can just be client/server timeouts. 

> State of connections for other keys than the one requested should not impact
> Not strictly required, but a good design sign: No IO should be done with locks held

For the outer locking, a more advanced 3rd party concurrent hashmap impl makes sense here to minimize that outer lock.

Reply:
I think we may have some terminology mismatch here, specifically around "full stream" .Here's my understanding, lmk if you disagree with anything and the names

A (outer) connection is a TCP/TLS connection 
On a connection we have HTTP 2 streams
1 stream and is 1 req+response pair

Each inner connection is 1:1 with a request (and thus 1:1 with a stream)

A typical pool will return a connection for a key. 
For our use cases, pool users are guaranteed to only send a single request on each usage - so we can specialize here, if desired, and expose an interface to send a single request/stream instead of connection level access

A connection is full of it has active streams equal to max streams 

a connection is idle if it has 0 streams on it 

---

If the pool doesn't close idle connection who else would? In theory the server could but this is racy (client should always have lower timeout to avoid sending on a closed connection) and insecure (we cannot trust someone else to manage our memory usage)

Reply:
> I think we may have some terminology mismatch here, specifically around "full stream" .Here's my understanding, lmk if you disagree with anything and the names

No, you're correct - I meant to type "full conn" and not "full stream".

>  For our use cases, pool users are guaranteed to only send a single request on each usage - so we can specialize here, if desired, and expose an interface to send a single request/stream instead of connection level access

Yeah - I considered that (use channels to send reqs and just expose that), but did not go down that road because it doesn't really solve any of the contention problems AFAICT. The pool still has to do pretty much the same locking/backpressure.

> If the pool doesn't close idle connection who else would? In theory the server could but this is racy (client should always have lower timeout to avoid sending on a closed connection) and insecure (we cannot trust someone else to manage our memory usage)

Probably several spots. Our server (in inbound.rs) should try to close idle connections. IIRC the `hyper` client should have its own idle timer, and we can just treat a closed client like a stream-maxed client, in terms of the pool - throw it out and start a new one.

Reply:
If you only remove when you try to use it then you will end up with zombie connections if they are not used anymore. The server closing it shouldn't be relied on (untrusted)

---

for the "1 request per checkout" I was thinking it might help track the active streams.

Reply:
> > If you only remove when you try to use it then you will end up with zombie connections if they are not used anymore. The server closing it shouldn't be relied on (untrusted)
> 
> I'm assuming both client and server can each independently configure close timeouts on their respective ends (and probably should).

Uh, I am not sure what happened to this comment, I was on mobile. I will revert it.

Reply:
> If you only remove when you try to use it then you will end up with zombie connections if they are not used anymore. The server closing it shouldn't be relied on (untrusted)

Yes, the client can't rely on the server doing it, and the server can't rely on the client doing it. So they both have to do it.

Clientside pool should ensure it drops references to clients that are closed, agreed.

Reply:
> Probably several spots, to be correct. Our server (in inbound.rs) should try to close idle connections. IIRC the hyper client should have its own idle timer, and we can just treat a closed client like a stream-maxed client, in terms of the pool - throw it out and start a new one, if we pull it out of the bag.

I agree with John that server-side close is likely to cause issues because it'll be racy. We should try to close on the client side if at all possible. Server-side closes should be last resort to manage resources where it's likely the client vanished without closing for some reason.

Reply:
> > Probably several spots, to be correct. Our server (in inbound.rs) should try to close idle connections. IIRC the hyper client should have its own idle timer, and we can just treat a closed client like a stream-maxed client, in terms of the pool - throw it out and start a new one, if we pull it out of the bag.
> 
> I agree with John that server-side close is likely to cause issues because it'll be racy. We should try to close on the client side if at all possible. Server-side closes should be last resort to manage resources where it's likely the client vanished without closing for some reason.

Agree - just saying we have to do server-side timeouts too, since that "last resort where the client vanishes without closing" is pretty easy to induce.

Reply:
ah, I misunderstood. yeah, we have to have some server-side time out but we need to be careful that it is strictly longer than all client timeouts.